aws/defaults/handlers.go

```diff
@@ -161,55 +162,84 @@ func handleSendError(r *aws.Request, err error) {
 
 	// Catch all request errors, and let the retryer determine
 	// if the error is retryable.
-	r.Error = awserr.New("RequestError", "send request failed", err)
+	r.Error = &aws.RequestSendError{Response: r.HTTPResponse, Err: err}
 
 	// Override the error with a context canceled error, if that was canceled.
 	ctx := r.Context()
 	select {
 	case <-ctx.Done():
-		r.Error = awserr.New(aws.ErrCodeRequestCanceled,
-			"request context canceled", ctx.Err())
-		r.Retryable = aws.Bool(false)
+		r.Error = &aws.RequestCanceledError{Err: ctx.Err()}
 	default:
 	}
 }
 
 // ValidateResponseHandler is a request handler to validate service response.
-var ValidateResponseHandler = aws.NamedHandler{Name: "core.ValidateResponseHandler", Fn: func(r *aws.Request) {
-	if r.HTTPResponse.StatusCode == 0 || r.HTTPResponse.StatusCode >= 300 {
-		// this may be replaced by an UnmarshalError handler
-		r.Error = awserr.New("UnknownError", "unknown error", nil)
-	}
-}}
+var ValidateResponseHandler = aws.NamedHandler{
+	Name: "core.ValidateResponseHandler",
+	Fn: func(r *aws.Request) {
+		if r.HTTPResponse.StatusCode >= 300 {
+			// This may be replaced by a protocol's UnmarshalError handler
+			r.Error = &aws.HTTPResponseError{Response: r.HTTPResponse}
+		}
+	}}
 
-// AfterRetryHandler performs final checks to determine if the request should
-// be retried and how long to delay.
-var AfterRetryHandler = aws.NamedHandler{
-	Name: "core.AfterRetryHandler",
+// RequestInvocationIDHeaderHandler sets the invocation id header for request
+// tracking across attempts.
+var RequestInvocationIDHeaderHandler = aws.NamedHandler{
+	Name: "core.RequestInvocationIDHeaderHandler",
 	Fn: func(r *aws.Request) {
-		// set retry state based on the service's state
-		r.Retryable = aws.Bool(r.Retryer.ShouldRetry(r))
+		const invocationIDHeader = "amz-sdk-invocation-id"
+		r.HTTPRequest.Header.Set(invocationIDHeader, r.InvocationID)
+	}}
 
-		if r.WillRetry() {
-			r.RetryDelay = r.Retryer.RetryRules(r)
+// RetryMetricHeaderHandler sets an additional header to the API request that
+// includes retry details for the service to consider.
+var RetryMetricHeaderHandler = aws.NamedHandler{
+	Name: "core.RetryMetricHeaderHandler",
+	Fn: func(r *aws.Request) {
+		const retryMetricHeader = "amz-sdk-request"
+		var parts []string
 
-			if err := sdk.SleepWithContext(r.Context(), r.RetryDelay); err != nil {
-				r.Error = awserr.New(aws.ErrCodeRequestCanceled,
-					"request context canceled", err)
-				r.Retryable = aws.Bool(false)
-				return
-			}
+		parts = append(parts, fmt.Sprintf("attempt=%d", r.AttemptNum))
+		if max := r.Retryer.MaxAttempts(); max != 0 {
+			parts = append(parts, fmt.Sprintf("max=%d", max))
+		}
+		// TODO "ttl=YYYYmmddTHHMMSSZ"
+		//    ttl = current_time + socket_read_timeout + estimated_skew
+		// SDK doesn't implement clock skew yet and not obvious how to obtain
+		// read deadlines.
```

Comment:
https://github.com/aws/aws-sdk-go-v2/issues/272 is the relevent issue to track this work, in case you want to throw that issue number in the PR